### ISSUE TITLE
Implement training function

### ISSUE THREAD

**Issue by jbirkesteen** [RasmusBrostroem/ConnectFourRL#10]
I've created one large issue instead of 3 small. If they aren't solved at the same time, we'll create new smaller issues as needed.
I've written them in what I think is the prioritised order.

## Train against itself
In order to not overfit the strategy against the random bot, we need to freeze the agent at some point and let it play against itself.
- need to implement a while-loop that prevents the freezed agent of playing illegal move.
_Possible issue:_ Consider whether this loop needs some sort of break-clause (if bot is 0.9999 percent certain it wants to stack illegal column, it could lead to a lot of repetitions). 
_Suggestion:_ Maybe just keep track of number of repetitions and choose a random legal column if the bot has chosen illegal column 10 times?

#### When do we want to freeze our agent?
Should this be based on passing some win-rate or just happen after some specified number of episodes?  
Will be interesting to see whether each new generation simply learns to stack-attack a new column, or if the agent learns to block after some number of generations.
**Idea for training:** What would happen if we let some agent train against a random generation for each batch_size of games?

## Save parameters of models
Easiest is probably to use `torch.save()`. We need some way of keeping track of the different models - either we should use something like NeptuneAI _(see slides for the last week with Kaare)_ or find a way of writing a log-document ourselves.  
How many models do we want to save? If we want to track how the strategy evolves, we probably want to save the first(or last) network in each generation, such that we can always see which strategy a given generation used.

## Modify hyperparameters
In order to make a well-informed choice of hyperparameters we need to be able to test (and track performance) of different settings. Again, consider whether it's simply easiest to begin using NeptuneAI for this.  
Ideally, we should be able to modify the following parameters (feel free to add more) through this function, but some (e.g. network architecture) might be difficult to change in this call?
- [x] batch_size (episodes before updating network)
- [x] network architecture
- [x] episodes before freezing network, or alternatively, win-rate required against opponent before freezing
- [x] learning_rate for optimizer
- [x] rewards for the possible outcomes

## Train strategy
Idea to train the agent in a more structured way:
- Train first few moves, like 8 or 10, and then punish the agent if no ending happend in these moves, together with win and lose reward
- Later train on more moves and so on

Another approach:
- Punish the agent for the number of moves it did, so more loss if it lost quick and less if it lost later, but more loss if it won late, and less loss if it won quick.

**Comment by RasmusBrostroem** [comment]
I agree on the issue in training against itself and the fact that we can get stuck in a while loop for the frozen agent. 
The idea of choosing a random move if the frozen agent wont choose a valid column seems fine to begin with. Furthermore, to begin with I think we should freeze our agent and start a new generation after a number of episodes, just to make it easier to compare different models, and the number of training epsiodes for each generation. After we can explore the win-rate possibility.

For the save parameters of models, I think we need to save the parameters in the end of a generation, so the next generation can play against the previous generation. I will look into torch save and load. 

For the hyperparameters, I will start our by just making it so we can decide on how we want to change it and then traing x number of generations, and then we can look into using NeptuneAI to better optimize the hyperparameters.

**Comment by jbirkesteen** [comment]
### Train against itself
Resolved with pr #18 - except the choice of when/how to freeze. This part of the topic is moved to discussions of hyperparameters (choice of `episodes_per_gen`)

### Saving parameters
Solved with #19
We upload all trained agents to our GitHub and create a text-file matching them with their corresponding NeptuneAI run IDs.
